libs/datasets/dataset_utils.py

```diff
@@ -279,3 +280,97 @@ def summarize(data, aggregate_level, groupby):
     print(key_fmt.format("Missing Fips Code:", missing_fips))
     print(key_fmt.format("Num non unique rows:", num_non_unique))
     print(index_size[index_size > 1])
+
+
+def fill_fields_with_data_source(
+    existing_df: pd.DataFrame,
+    data_source: pd.DataFrame,
+    index_fields: List[str],
+    columns_to_fill: List[str],
+) -> pd.DataFrame:
+    """Pull columns from an existing data source into an existing data frame.
+
+    Example:
+
+        existing_df:
+        ----------------
+        | date | cases |
+        | 4/2  | 1     |
+        | 4/3  | 2     |
+        | 4/4  | 3     |
+        ----------------
+
+        data_source:
+        ----------------------
+        | date | current_icu |
+        | 4/3  | 4           |
+        | 4/5  | 5           |
+        ----------------------
+
+        index_fields: ['date']
+
+        columns_to_fill: ['current_icu']
+
+        output:
+        ------------------------------
+        | date | cases | current_icu |
+        | 4/2  | 1     | Na          |
+        | 4/3  | 2     | 4           |
+        | 4/4  | 3     | Na          |
+        | 4/5  | Na    | 5           |
+        ------------------------------
+
+    Args:
+        existing_df: Existing data frame
+        data_source: Data used to fill existing df columns
+        index_fields: List of columns to use as common index.
+        columns_to_fill: List of columns to add into existing_df from data_source
+
+    Returns: Updated dataframe with requested columns filled from data_source data.
+    """
+    new_data = data_source.set_index(index_fields)
+
+    # If no data exists, return all rows from new data with just the requested columns.
+    if not len(existing_df):
+        for column in columns_to_fill:
+            if column not in new_data.columns:
+                new_data[column] = None
+        return new_data[columns_to_fill].reset_index()
+    existing_df = existing_df.set_index(index_fields)
+
+    # Sort indices so that we have chunks of equal length in the
+    # correct order so that we can splice in values.
+    existing_df = existing_df.sort_index()
+    new_data = new_data.sort_index()
+
+    # Build series that point to rows that match in each data frame.
+    existing_df_in_new_data = existing_df.index.isin(new_data.index)
+    new_data_in_existing_df = new_data.index.isin(existing_df.index)
```

Comment:
the section above feels like it really should be done via a join, no?

Reply:
I found the join tricky with only updating a few columns. would be down to do it with a join if it's possible.